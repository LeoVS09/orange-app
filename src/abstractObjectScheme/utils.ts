import { AosField, AosFieldType, AosSimpleField, AosRelationsField } from './types';

export const isSimpleAosField = (field: AosField): field is AosSimpleField => 
    !isRelationsAosField(field)

export const isRelationsAosField = (field: AosField): field is AosRelationsField => 
    field.type === AosFieldType.OneToOne || 
    field.type === AosFieldType.OneToMany
    
