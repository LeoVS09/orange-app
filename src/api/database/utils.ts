import {APIClient, makeClient} from './client'
import urls from '../urls.json'
import deepMap from 'deep-map'

// const DEBUG = process.env.NODE_ENV !== 'production'
const DEBUG = false

export const client = makeClient(urls.DATABASE_SERVER)


// TODO: complete typesation for non parametred queries
export function generateQuery<V, T, R>(query: any, formatter: (result: T) => R | null, mock?: (variables: V) => Promise<R>) {
   if (DEBUG && mock) {
      return (variables: V) => mock(variables)
   }

   return mapRequester((client) => (variables: V) =>
      client.query<T>({
         query,
         variables,
      })
         .then((result) => result.data && formatter(result.data)))
}

export function generateSimpleQuery<T, R>(query: any, formatter: (result: T) => R | null, mock?: () => Promise<R>) {
   if (DEBUG && mock) {
      return () => mock()
   }

   return mapRequester((client) => () =>
      client.query<T>({
         query,
      })
         .then((result) => result.data && formatter(result.data)),
   )
}

export function generateMutation<V, T, R>(mutation: any, formatter: (result: T) => R | null, mock?: (variables: V) => Promise<R>) {
   if (DEBUG && mock) {
      return (variables: V) => mock(variables)
   }

   return mapRequester((client) => (variables: V) =>
      client.mutate<T>({
         mutation,
         variables,
      })
         .then((result) => result.data && formatter(result.data)),
   )
}

// mapRequester transform date string to date objects, but currently TypeScript don't have support
// for so advanced templates, so just know this fact when set types of request result

function mapRequester<R>(requester: (client: APIClient) => R): R  {
   const wrapped = requester(client)

   // @ts-ignore
   return async (input: any) => {
      // @ts-ignore
      const result = await wrapped(input)
      if (!result) {
         return result
      }

      return dateToStringFormatter(result)
   }
}

const keysForMapDays: string[] = []
const keyMath = /.+(Date|At)$/gm

// Transform to date types fields with matched names
function dateToStringFormatter <T>(t: T): T {
   return deepMap<T>(t, (value, key) => {
      if (typeof key === 'number') {
         return value
      }

      if (keysForMapDays.indexOf(key) === -1 && !key.match(keyMath)) {
         return value
      }

      if (value === null || value === undefined) {
         return value
      }

      return new Date(value)
   })
}
