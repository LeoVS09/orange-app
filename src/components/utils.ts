
export function toStringWhenDefined(value: any): string {
  if (value === undefined)
    return ''

  if (value === null)
    return ''

  return `${value}`
}

export function isDate(date: any) {
  return Object.prototype.toString.call(date) === '[object Date]'
}

export function formatDate(date: Date) {
  if (!date)
    return ''

  return date.toLocaleDateString()
}

export function randomId(): string {
  return Math.random().toString(36).substring(2) + Math.random().toString(36).substring(2)
}

export interface IOnWheelCallbacks {
   up?: () => void
   down?: () => void
}

export function onWheel(elem: Element, { up = () => { }, down = () => { } }: IOnWheelCallbacks) {
  const onWheelListener: EventListener = e => {
    e = e || window.event

    // wheelDelta не дает возможность узнать количество пикселей
    // @ts-ignore
    const delta = e.deltaY || e.detail || -e.wheelDelta

    if (e.preventDefault)
      e.preventDefault()
    else
      e.returnValue = false

    if (delta < 0)
      up()
    else
      down()
  }

  if (elem.addEventListener) {
    if ('onwheel' in document) {
      // IE9+, FF17+, Ch31+
      elem.addEventListener('wheel', onWheelListener)
    } else if ('onmousewheel' in document) {
      // устаревший вариант события
      elem.addEventListener('mousewheel', onWheelListener)
    } else {
      // Firefox < 17
      elem.addEventListener('MozMousePixelScroll', onWheelListener)
    }
  } else {
    // IE8-
    // @ts-ignore
    elem.attachEvent('onmousewheel', onWheelListener)
  }
}
