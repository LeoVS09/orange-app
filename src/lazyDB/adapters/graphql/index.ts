import { AosSchema } from '@/abstractObjectSchema'
import { GenerateQueryOptions, generateQuery } from './generateQuery'
import { schemaToQueryFields } from './schemaToQuery'

export * from './types'
export * from './schemaToQuery'
export * from './generateQuery'

export interface GenerateQueryFromAosSchemaOptions extends Omit<GenerateQueryOptions, 'fields'> {
    /** fields which will be inside query */
    schema: AosSchema
  }

/** Generate query from AOS schema, for more info check generateQuery function */
export const generateQueryFromAosSchema = ({ operation, schema, variables }: GenerateQueryFromAosSchemaOptions) =>
  generateQuery({
    operation,
    fields: schemaToQueryFields(schema),
    variables
  })
