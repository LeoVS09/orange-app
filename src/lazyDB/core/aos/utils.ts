
const ANY_NON_DIGIT_CHARACTER = /\D/

// TODO: need find better solution for resolve issue

/**
 * Will transform string to number if it possible,
 * actually resolve problem: "When asked item in array by index, index number transform to string",
 * as example: array[3] -> will by catched by proxy as get("3").
 * For resolve this issue, allways tryies cast any string to number.
 * Actaully cannot be casted only for array, because js, under the hood any array actaully is object,
 * and any object with number keys can be interpritated as array. (V8 work this way)
 */
export function toNumberIfCan(name: string | number | symbol): string | number {
  // Symbol just for correct ts typing
  if (typeof name === 'number' || typeof name === 'symbol')
    return name as number

  // Prevent to incorrectly parse id as number
  if (ANY_NON_DIGIT_CHARACTER.test(name))
    return name

  const parsedNumber = Number.parseInt(name, 10)
  if (Number.isInteger(parsedNumber))
    return parsedNumber

  return name
}
