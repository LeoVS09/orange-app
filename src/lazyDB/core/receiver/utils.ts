
export const isPromiseLike = (value: any): value is PromiseLike<any> =>
  typeof value === 'object' && value.then

// TODO: rewrite, code must write without this strange behavior
export function async<T>(value: T | Promise<T>): Promise<T> {
  if (isPromiseLike(value))
    return value

  return Promise.resolve(value)
}
