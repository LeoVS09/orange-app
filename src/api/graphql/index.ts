import * as mutations from './mutations'
import * as queries from './queries'
import {mapRequester} from "./utils";

// TODO: add pre-compiling queries when build project

export const login = mapRequester(mutations.login);

export const register = mapRequester(mutations.register);

export const currentUser = mapRequester(queries.currentUser);

export const searchCountries = mapRequester(queries.searchCountries);

export const countries = mapRequester(queries.countries);

export const problems = mapRequester(queries.problems)

export const problem = mapRequester(queries.problem)
