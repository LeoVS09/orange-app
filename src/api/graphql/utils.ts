import {APIClient, makeClient} from "../graphql/apollo";
import urls from "../urls.json";
import deepMap from "deep-map";

const client = makeClient(urls.DATABASE_SERVER);

// mapRequester transform date string to date objects, but currently TypeScript don't have support
// for so advanced templates, so just know this fact when set types of request result

export function mapRequester<R>(requester: (client: APIClient) => R): R  {
   const wrapped = requester(client)

   // @ts-ignore
   return async (input: any) => {
      // @ts-ignore
      const result = await wrapped(input)
      if(!result)
         return result

      return dateToStringFormatter(result)
   }
}

const keysForMapDays = ['createdAt', 'updatedAt', 'publishedAt', 'publicationDate']

// Transform to date types fields with matched names
function dateToStringFormatter <T>(t: T): T {
   return deepMap<T>(t, (value, key) => {
      if(typeof key === "number")
         return value

      if(keysForMapDays.indexOf(key) === -1)
         return value

      return new Date(value)
   })
}
