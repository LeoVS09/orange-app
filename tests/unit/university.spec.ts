// @ts-ignore next-line
global.fetch = require('node-fetch')
import nock from 'nock'
import { createLocalVue, mount } from '@vue/test-utils'
import { VueConstructor } from 'vue'
import Vuex, {Store} from 'vuex'
import { nockAllGraphqlRequests } from './utils/graphql.mock'
import cons from './utils/console.mock'
import { WaitStore } from './utils/wait'
import { when } from './utils/when'
import vuexI18n from 'vuex-i18n'

describe('University.vue', () => {

  let localVue: VueConstructor<any> 
  let store: Store<any> 
  let requests: WaitStore

  beforeAll(() => nock.disableNetConnect());

  beforeEach(async () => {
    localVue = createLocalVue()
    localVue.use(Vuex)
    store = new Store({
      getters: {
        isTeacher: () => false
      }
    })

    localVue.use(vuexI18n.plugin, store)

    const { waits } = await nockAllGraphqlRequests(nock, 'university')
    requests = waits
    cons.mockConsole()
  })

  afterEach(() => {
    cons.restoreConsole()
    nock.cleanAll()
  });

  it('renders university', async () => {
    // must be inside test for mock fetch function
    const Component = require('@/pages/University.vue').default
    
    const wrapper = mount(Component, { 
      store, 
      localVue, 
      attachToDocument: true,
      propsData: { 
        id: 'university-test-id'
      }
     })
    
    expect(wrapper.element).toMatchSnapshot()

    expect(wrapper.text()).not.toContain('numquam')
    
    // Wait while lazyDB requested all data and rerender
    await requests.wait()
    await when(() => wrapper.text().includes('numquam'))

    expect(wrapper.text()).toContain('numquam')
    expect(wrapper.text()).toContain('Cook Islands')
    expect(wrapper.text()).toContain('West Devanstad')

    expect(wrapper.element).toMatchSnapshot()
  })
})
