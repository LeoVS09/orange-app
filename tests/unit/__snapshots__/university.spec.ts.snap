// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`University.vue renders university 1`] = `
<div
  class="university"
>
  <div
    class="page-header highlight"
  >
    <div
      class="page-header--hidden-breadcrumbs"
    >
      <div
        class="breadcrumbs"
      >
        Countries
      </div>
       
      <div
        class="breadcrumbs"
      >
        Country
      </div>
       
      <div
        class="breadcrumbs"
      >
        City
      </div>
    </div>
     
    <!---->
     
    <div
      class="page-header--content"
    >
      <h1>
        
      </h1>
       
      <div
        class="page-header--actions"
      />
       
      <!---->
    </div>
     
    <!---->
  </div>
   
  <section
    class="lazy-data section without-margin border-bottom"
  >
    <div
      class="lazy-data--main"
    >
      <h4
        class="lazy-data--header"
      >
        Information
      </h4>
       
      <div
        class="lazy-data--hidden-columns"
      >
        <div
          class="model-property"
        >
          longName
        </div>
      </div>
       
      <div
        class="lazy-data--content"
      />
    </div>
  </section>
</div>
`;

exports[`University.vue renders university 2`] = `
<div
  class="university"
>
  <div
    class="page-header highlight breadcrumbs"
  >
    <div
      class="page-header--hidden-breadcrumbs"
    >
      <div
        class="breadcrumbs"
      >
        Countries
      </div>
       
      <div
        class="breadcrumbs"
      >
        Cook Islands
      </div>
       
      <div
        class="breadcrumbs"
      >
        West Devanstad
      </div>
    </div>
     
    <div
      class="page-header--breadcrumbs"
    >
      <router-link
        class="page-header--breadcrumb-step link"
        to="[object Object]"
      >
        Countries
      </router-link>
       
      <span
        class="page-header--breadcrumb-separator"
      >
        /
      </span>
      <router-link
        class="page-header--breadcrumb-step link"
        to="[object Object]"
      >
        Cook Islands
      </router-link>
       
      <span
        class="page-header--breadcrumb-separator"
      >
        /
      </span>
      <router-link
        class="page-header--breadcrumb-step link"
        to="[object Object]"
      >
        West Devanstad
      </router-link>
       
      <span
        class="page-header--breadcrumb-separator"
      >
        /
      </span>
    </div>
     
    <div
      class="page-header--content"
    >
      <h1>
        numquam
      </h1>
       
      <div
        class="page-header--actions"
      />
       
      <!---->
    </div>
     
    <div
      class="page-header--data"
    >
      <div
        class="page-header--data-content"
      >
        <div
          class="data-view in-row contrast small"
        >
          <div
            class="data-view--item"
          >
            <i
              class="material-icons data-view--icon"
            >
              publish
            </i>
             
            <span
              class="data-view--value"
            >
              25.05.2022
            </span>
          </div>
          <div
            class="data-view--item"
          >
            <i
              class="material-icons data-view--icon"
            >
              edit
            </i>
             
            <span
              class="data-view--value"
            >
              17.04.2020
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
   
  <section
    class="lazy-data section without-margin border-bottom"
  >
    <div
      class="lazy-data--main"
    >
      <h4
        class="lazy-data--header"
      >
        Information
      </h4>
       
      <div
        class="lazy-data--hidden-columns"
      >
        <div
          class="model-property"
        >
          longName
        </div>
      </div>
       
      <div
        class="lazy-data--content"
      />
    </div>
  </section>
</div>
`;
